editor/src/components/canvas/controls/elements-outside-visible-area-hooks.tsx

```diff
@@ -0,0 +1,308 @@
+    return {
+      x: bounds.x * scaleRatio,
+      y: bounds.y * scaleRatio,
+      width: bounds.width * scaleRatio - navigatorWidth,
+      height: bounds.height * scaleRatio,
+    } as WindowRectangle
+  }, [bounds, navigatorWidth, canvasScale])
+
+  const scaledCanvasAreaCenter = React.useMemo(() => {
+    if (scaledCanvasArea == null) {
+      return null
+    }
+    return getRectCenter(scaledCanvasArea)
+  }, [scaledCanvasArea])
+
+  const elementsOutsideVisibleArea = React.useMemo(() => {
+    return mapDropNulls((path: ElementPath): ElementOutsideVisibleArea | null => {
+      if (scaledCanvasArea == null) {
+        return null
+      }
+      const frame = framesByPathString[EP.toString(path)]
+      if (frame == null) {
+        return null
+      }
+
+      const topLeftSkew = { x: -navigatorWidth, y: 0 } as WindowPoint
+      const topLeftPoint = offsetPoint(
+        canvasPointToWindowPoint(frame, canvasScale, canvasOffset),
+        topLeftSkew,
+      )
+      const elementRect = {
+        x: topLeftPoint.x,
+        y: topLeftPoint.y,
+        width: frame.width * canvasScale,
+        height: frame.height * canvasScale,
+      } as WindowRectangle
+
+      const directions = getOutsideDirections(scaledCanvasArea, elementRect)
+      if (directions.length === 0) {
+        return null
+      }
+
+      return {
+        path: path,
+        rect: elementRect,
+        directions: directions,
+      }
+    }, elements)
+  }, [elements, canvasOffset, canvasScale, scaledCanvasArea, framesByPathString, navigatorWidth])
+
+  return React.useMemo((): ElementOutsideVisibleAreaIndicator[] => {
+    if (
+      scaledCanvasArea == null ||
+      scaledCanvasAreaCenter == null ||
+      bounds == null ||
+      canvasToolbar == null
+    ) {
+      return []
+    }
+
+    return (
+      elementsOutsideVisibleArea
+        // Map elements to indicators
+        .map(({ rect, path, directions }): ElementOutsideVisibleAreaIndicator => {
+          return {
+            id: getIndicatorId(path, directions),
+            path: path,
+            cluster: 1,
+            angle: angleBetweenPoints(scaledCanvasAreaCenter, getRectCenter(rect)),
+            position: adjustPosition(
+              offsetPoint(rect, {
+                x: rect.width / 2,
+                y: rect.height / 2,
+              } as WindowPoint),
+              directions,
+              scaledCanvasArea,
+              navigatorWidth,
+              windowRectangleFromDOMRect(canvasToolbar),
+            ),
+          }
+        })
+        // Group the indicators into clusters
+        .reduce((arr, indicator): ElementOutsideVisibleAreaIndicator[] => {
+          const index = arr.findIndex((other) => {
+            const distanceBetween = distance(indicator.position, other.position)
+            return distanceBetween < minClusterDistance
+          })
+          if (index >= 0) {
+            arr[index].cluster++
```

Comment:
If this was a loop, then you could also eliminate the intermediary array that the above `.map` produces.

Reply:
done in https://github.com/concrete-utopia/utopia/pull/3842/commits/4de4fe65103e0428b3f209f33331295de2f91ca7